### PR TITLE
fix(sysinfo): address permission errors in Base.Sys.which

### LINKED ISSUE #49181: `Base.Sys.which` errors if any directory without read access is present in PATH
The `Base.Sys.which` is intended to mimic the `which` tool on Unix, but it errors if there is any folder in the `PATH` variable that does not have read privilege. The Unix `which` tool does not seem to have this behaviour.

I noticed this in an HPC environment where the administrator has included some directories in the PATH where a normal use does not have access. Several tools seems to be relay on `Base.Sys.which`, such as PackageCompiler and `p7zip_jll.jl` in the stdlib

Minimum non-working example:

```bash
mkdir noaccess
chmod 000 noaccess
which echo
export PATH=`pwd`/noaccess:$PATH
julia -e 'Base.Sys.which("echo")'
```

output 

```
ERROR: IOError: stat("/home/bonan/tmp/noaccess/echo"): permission denied (EACCES)
Stacktrace:
 [1] uv_error
   @ ./libuv.jl:97 [inlined]
 [2] stat(path::String)
   @ Base.Filesystem ./stat.jl:152
 [3] isfile
   @ ./stat.jl:456 [inlined]
 [4] which(program_name::String)
   @ Base.Sys ./sysinfo.jl:519
 [5] top-level scope
   @ none:1

```

The issue is due to the use of `isfile` call:

https://github.com/JuliaLang/julia/blob/afb6c60d69a38e8a2442a0c7e87c47b8880ad294/base/sysinfo.jl#L519-L520

No sure what is the best solution in this case, but I think `which` should not error if there is inaccessible folder. 
One solution may to include a try...catch block. 

My versioninfo():

```
Commit afb6c60d69a (2022-09-06 15:09 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 20 ├ù 12th Gen Intel(R) Core(TM) i7-12700H
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, goldmont)
  Threads: 1 on 20 virtual cores
Environment:
  LD_LIBRARY_PATH = /home/bonan/appdir/n2p2/lib:/home/bonan/.local/libxc/lib:/home/bonan/.local/fftw3/lib:/home/bonan/appdir/n2p2/lib:/home/bonan/.local/libxc/lib:/home/bonan/.local/fftw3/lib:/opt/intel/oneapi/tbb/2021.8.0/env/../lib/intel64/gcc4.8:/opt/intel/oneapi/mkl/latest/lib/intel64:/opt/intel/oneapi/compiler/2023.0.0/linux/lib:/opt/intel/oneapi/compiler/2023.0.0/linux/lib/x64:/opt/intel/oneapi/compiler/2023.0.0/linux/compiler/lib/intel64_lin
```

### PR DESCRIPTION
**Pull request by wldeh**
Closes #49181. Rectifies the discrepancy between the behavior of Base.Sys.which in Julia and the Unix which command when encountering directories without read access in the PATH environment variable.

### CODE REVIEW

**Comment by Seelengrab** [comment]
Closes https://github.com/JuliaLang/julia/issues/49181

**Comment by wldeh** [comment]
> Closes #49181 - might want to add that to the OP, so Github links the issue & the PR :)

Thank you @Seelengrab for pointing that out.

File: base/sysinfo.jl [hunk 1]
```diff
@@ -543,9 +543,20 @@ function which(program_name::String)
...
-                return program_path
+            try
+                # If we find something that matches our name and we can execute
+                if isfile(program_path) && isexecutable(program_path)
+                    return program_path
+                end
+            catch e
+                # If we encounter a permission error, we skip this directory
+                # and continue to the next directory in the PATH variable.
+                if isa(e, IOError) && e.code == Base.UV_EACCES
```
Comment (staticfloat) [inline base/sysinfo.jl:554]: Is there some way to create a test to exercise this code path? Maybe we create two temporary directories and put “foo” inside them both, give one directory zero permissions, add it to the front of PATH, then ensure that Sys.which(“foo”) works as expected.

**Review by staticfloat — CHANGES_REQUESTED** [review]

File: base/sysinfo.jl (hunk 1 above)
Comment (vtjnash) [reply base/sysinfo.jl:554]: This should use the `access` syscall, not the `isexecutable` metadata (which, despite the name, doesn't tell you if you can execute it)

Comment (staticfloat) [reply base/sysinfo.jl:554]: I think the issue is this:
```
julia> Sys.isexecutable(mktempdir())
true
```

Comment (giordano) [reply base/sysinfo.jl:554]: #49181 has a reproducer which is a mix of shell commands and julia code, but it shouldn't be hard to write everything in Julia.

**Comment by wldeh** [comment]
Apologies for the delay! Added tests based on the [comment from staticfloat](https://github.com/JuliaLang/julia/pull/49229#discussion_r1156029187).

**Review by staticfloat — APPROVED** [review]

**Pull request merged**